File: app/src/main/java/com/bitcoin/merchant/app/network/websocket/impl/bitcoincom/BitcoinComSocketHandler.java
package com.bitcoin.merchant.app.network.websocket.impl.bitcoincom;

import android.util.Log;

import com.bitcoin.merchant.app.model.PaymentReceived;
import com.bitcoin.merchant.app.model.Tx;
import com.bitcoin.merchant.app.network.ExpectedAmounts;
import com.bitcoin.merchant.app.network.ExpectedPayments;
import com.bitcoin.merchant.app.network.websocket.impl.TxWebSocketHandlerImpl;
import com.neovisionaries.ws.client.WebSocket;
import com.neovisionaries.ws.client.WebSocketFactory;

import org.bitcoindotcom.bchprocessor.bip70.GsonHelper;

import java.io.IOException;

public class BitcoinComSocketHandler extends TxWebSocketHandlerImpl {
    public BitcoinComSocketHandler() {
        TAG = "BitcoinComSocket";
    }

    @Override
    protected WebSocket createWebSocket(WebSocketFactory factory) throws IOException {
        return factory.createSocket("wss://bch.api.wallet.bitcoin.com/bws/api/socket/v1/address");
    }

    @Override
    protected void parseTx(String message) throws Exception {
        Tx tx = GsonHelper.INSTANCE.getGson().fromJson(message, Tx.class);
        if (tx != null && tx.outputs != null) {
            Log.i(TAG, "TX found:" + tx);
            for (Tx.Output o : tx.outputs) {
                String addr = o.address;
                if (ExpectedPayments.getInstance().isValidAddress(addr)) {
                    ExpectedAmounts expected = ExpectedPayments.getInstance().getExpectedAmounts(addr);
                    long bchReceived = o.value;
                    if (webSocketListener != null) {
                        long timeInSec = System.currentTimeMillis() / 1000;
                        PaymentReceived payment = new PaymentReceived(addr, bchReceived, tx.txid, timeInSec, 0, expected);
                        Log.i(TAG, "expected payment:" + payment);
                        webSocketListener.onIncomingPayment(payment);
                    }
                }
            }
        }
    }
}

File: app/src/main/java/com/bitcoin/merchant/app/network/websocket/impl/bitcoincom/BitcoinComSocketHandler.kt
package com.bitcoin.merchant.app.network.websocket.impl.bitcoincom

import android.util.Log
import com.bitcoin.merchant.app.model.PaymentReceived
import com.bitcoin.merchant.app.model.Tx
import com.bitcoin.merchant.app.network.ExpectedAmounts
import com.bitcoin.merchant.app.network.ExpectedPayments
import com.bitcoin.merchant.app.network.websocket.impl.TxWebSocketHandlerImpl
import com.neovisionaries.ws.client.WebSocket
import com.neovisionaries.ws.client.WebSocketFactory
import org.bitcoindotcom.bchprocessor.bip70.GsonHelper.gson
import java.io.IOException

class BitcoinComSocketHandler : TxWebSocketHandlerImpl() {
    @Throws(IOException::class)
    override fun createWebSocket(factory: WebSocketFactory): WebSocket {
        return factory.createSocket("wss://bch.api.wallet.bitcoin.com/bws/api/socket/v1/address")
    }

    @Throws(Exception::class)
    override fun parseTx(message: String?) {
        val tx = gson.fromJson(message, Tx::class.java)
        if (tx != null && tx.outputs != null) {
            Log.i(TAG, "TX found:$tx")
            for (o in tx.outputs!!) {
                val addr = o.address
                if (ExpectedPayments.instance.isValidAddress(addr)) {
                    val expected: ExpectedAmounts = ExpectedPayments.instance.getExpectedAmounts(addr)
                    val bchReceived = o.value
                    if (webSocketListener != null) {
                        val timeInSec = System.currentTimeMillis() / 1000
                        val payment = PaymentReceived(addr, bchReceived, tx.txid, timeInSec, 0, expected)
                        Log.i(TAG, "expected payment:$payment")
                        webSocketListener!!.onIncomingPayment(payment)
                    }
                }
            }
        }
    }

    init {
        TAG = "BitcoinComSocket"
    }
}